Add bindings to std::function & a workaround for calling it in a timer

AndrejMitrovic opened the pull request:
The API change might make it more difficult to apply new SCP change-sets from upstream.

I've tried to find a way to increment the refcount for `std::function` with a bit of googling, but with no luck. Maybe there is a better way to do this though, I'm open to suggestions.

Geod24 commented on source/scpd/Cpp.d:
```diff
@@ -111,6 +111,16 @@ extern(C++, `std`) {
     }
+
+    /// Simple bindings to std::chrono
+    public struct chrono
```
Isn't `chrono` a namespace ??

AndrejMitrovic replied:
Right, I just wanted the mangling to match.

Geod24 commented on source/scpd/Cpp.d:
```diff
@@ -111,6 +111,16 @@ extern(C++, `std`) {
+    /// Simple bindings to std::chrono
+    public struct chrono
+    {
+        public static struct duration
```
Isn't it a template class ?

AndrejMitrovic replied:
Hmm.. So I looked into the sources. To make `duration` templated I need to turn the struct into:

```D
public static struct duration (_Rep, _Period = ratio!1)
```

And then also bind to `ratio`, which is another templated struct:

```C++
namespace std _GLIBCXX_VISIBILITY(default)
{
    template<intmax_t _Num, intmax_t _Den = 1>
        struct ratio
        {
        }
}
```

Is it ok to bind to the `_GLIBCXX_VISIBILITY` namespace? Because it's specific to glibc and our D binding code might break on other platforms..

Geod24 replied:
The namespace is `std`, `_GLIBCXX_VISIBILITY` is just a macro (which does nothing in this case).

Geod24 reviewed (commented):
Hum, I'll need a bit more time to review this. So the timers are not recurring, they are one-shot timers, right ?

AndrejMitrovic commented:
> Hum, I'll need a bit more time to review this. 

No rush. This will have to carry-over to the next sprint.

> So the timers are not recurring, they are one-shot timers, right ?

They seem to be from what I can tell.

Geod24 reviewed (changes requested):
- "Use workaround ..." commit LGTM
- Pending review on `std::chrono` binding
- Can you rebase with `--reset-author` ? Github f'ed the order, as usual

AndrejMitrovic committed:
Implement simple bindings for std::function

Can't call it 'function' due to reserved keyword.

AndrejMitrovic committed:
Use workaround for caling std::function delegates

Previously, the SCP C++ code allocated an std::function
which would be collected at the end of the function scope
because the reference count remained at 1.

To work around this, the delegate is now assigned to
a pointer allocated on the heap.

Once the delegate is invoked, it's then destroyed
as it will only be called once.

The setupTimer() API was changed to take a pointer to
an std::function, rather than pass it by value.

The callCPPDelegate() wrapper was introduced,
which calls the provided std::function*,
and deletes the pointer after the function call.

AndrejMitrovic committed:
Implement bindings to std::chrono::duration

AndrejMitrovic commented:
Aha, you want me to make std.chrono a namespace?

Geod24 commented:
> Aha, you want me to make std.chrono a namespace?

Yep, or is there an issue I'm not seeing ?

AndrejMitrovic commented:
Fixed the bindings.
